### PR TITLE
WIP: updating Schema and system tables

### PR DESCRIPTION
**Pull request by kazimuth**
# Description of Changes

Sketch of updating the system tables + updating ModuleDef to be a little easier to add constraints to, as suggested by Mazdak a while back.

I will probably split off the ModuleDef update into its own PR, since it is fairly contained and can be merged in the next day or two.
That is blocked on merging https://github.com/clockworklabs/SpacetimeDB/pull/1658 and https://github.com/clockworklabs/SpacetimeDB/pull/1661.

This code is full of errors, the draft is just here to allow review of the new Schema definitions.

# API and ABI breaking changes

This breaks the ABI and API, it rearranges system tables.

# Expected complexity level and risk

4, this is going to reverberate across the entire codebase.

# Testing

*Describe any testing you've done, and any testing you'd like your reviewers to do,
so that you're confident that all the changes work as expected!*

- [x] *Write a test you've completed here.*
- [ ] *Write a test you want a reviewer to do here, so they can check it off when they're satisfied.*

### CODE REVIEW

**Comment by kazimuth** [comment]
Key questions: 
- Can the system tables store enums (with data) at all?
- Can we add variants to an enum stored in the system tables without it being a breaking change?
- Can we add a new system table without it being a breaking change?

**Comment by gefjon** [comment]
First let me say I'm addressing this from the perspective of: we maintain a known schema for each system table, supplied by the host, and the data in the snapshot and/or log must conform to that schema. That may not always be the case, but it is the case now, and frankly I don't think we have time to change it before 1.0.

> Can the system tables store enums (with data) at all?

Sure. Why wouldn't they be able to?

> Can we add variants to an enum stored in the system tables without it being a breaking change?

First of all, obviously you can't take a newer module, commitlog or snapshot and run it on an older host if you do that. The old host will see what it believes to be an invalid discriminant and likely panic.

Second of all, you can only take an older module, commitlog or snapshot and run it on a newer host if the new variant doesn't change the BFLATN layout of the sum type. This means that the new variant's payload must have size and alignment less than or equal to the largest and most-aligned previously-existing variant(s).

> Can we add a new system table without it being a breaking change?

Again only in the new-host-old-data sense, and not the old-host-new-data sense, but sure, I don't see why not. I would expect a not-present table to behave the same as an empty table in either a snapshot or a commitlog. 'Course, we should test that, because there's a good chance I made an incorrect assumption somewhere in the snapshot-restoration code, but I don't think there's a fundamental problem or even a significant barrier to this working.

File: crates/core/src/db/datastore/system_tables.rs [hunk 1]
```diff
@@ -208,13 +212,16 @@ st_fields_enum!(
...
+st_fields_enum!(
+    enum StConstraintFields {
+        "constraint_id", ConstraintId = 0,
+        "name", ConstraintName = 1,
+        // TODO(jgilles): can a system table store an enum?
+        // What if we need to add new variants to the enum? Will that break
+        // the schema? Does this need to store a byte array instead?
+        "data", ConstraintData = 3,
+    }
+);
```
Comment (cloutiertyler) [inline crates/core/src/db/datastore/system_tables.rs:224]: I would expect this to mirror how constraints are represented in `ModuleDef`. If we're going to have an enum here, that's what we should do in `ModuleDef` as well.

Notably postgres appears to put them all in a single table, but I would like you to confirm that.

Comment (cloutiertyler) [reply crates/core/src/db/datastore/system_tables.rs:224]: Ah JK you seem to have done that below.

File: crates/core/src/db/datastore/system_tables.rs [hunk 2]
```diff
@@ -173,26 +173,30 @@ macro_rules! st_fields_enum {
 // WARNING: For a stable schema, don't change the field names and discriminants.
 st_fields_enum!(enum StTableFields {
     "table_id", TableId = 0,
-    "table_name", TableName = 1,
+    "name", TableName = 1,
     "table_type", TableType = 2,
     "table_access", TablesAccess = 3,
 });
```
Comment (cloutiertyler) [inline crates/core/src/db/datastore/system_tables.rs:179]: If you're going to rename `table_name` to name, might as well do `table_type` and `table_access`.

Incidentally, I don't know what `table_type` even is. 

Or better yet, I would just leave these names as is to reducer code churn and errors and then we can do a final pass at the end. So leave it as:

```rust
st_fields_enum!(enum StTableFields {
    "table_id", TableId = 0,
    "table_name", TableName = 1,
    "table_name", TableName = 1,
    "table_type", TableType = 2,
    "table_access", TablesAccess = 3,
});
```

File: crates/core/src/db/datastore/system_tables.rs [hunk 3]
```diff
@@ -173,26 +173,30 @@ macro_rules! st_fields_enum {
...
+st_fields_enum!(
+    enum StIndexFields {
+        "index_id", IndexId = 0,
+        "table_id", TableId = 1,
+        "index_name", IndexName = 2,
+
+        // TODO(jgilles): can a system table store an enum?
+        // What if we need to add new variants to the enum? Will that break
+        // the schema? Does this need to store a byte array instead?
+        "index_algorithm", IndexAlgorithm = 3,
```
Comment (cloutiertyler) [inline crates/core/src/db/datastore/system_tables.rs:197]: This unfortunately doesn't tell me what the type of this field is.

Could you put a schema for all the tables in a comment like this:

```md
| table_id | table_name  | table_type | foobar          |
|----------|-------------|------------|-----------------|
| u32      | String      | u32        | IndexAlgorithm  |

enum IndexAlgorithm {
    Foobar(String)
}
```

Or something similar? It's just hard to comment on the changes without the type info.

**Comment by kazimuth** [comment]
One other questions from Tyler:
> Should we store primary_key in the schema? I realize we don't currently use it and that it's not meaningful except to determine what should be an update vs an insert/delete, but how do you compare two ModuleDefs which differ by PrimaryKey in ponderautomigrate? I guess it just ignores them?

`primary_key` can always be extracted by loading the `ModuleDef`. I guess we need to decide if you need to be able to exactly reconstruct the `ModuleDef` from the system tables. Choices:
1. Keep information that only matters to client codegen in `ModuleDef`, not in the system tables.
    We'll need to keep `ModuleDef`s around next to the system tables for performing `ponder_automigrate`.
2. Add client codegen information (Type names, reducer names, misc exports, the typespace) to the system tables.
    Then we can exactly reconstruct the input `ModuleDef` as needed.

I'm not sure which is the best option here. 2 leaks some "nominal" information into the system tables, on the other hand, 1 stores information redundantly. 🤷

File: crates/core/src/db/datastore/system_tables.rs [hunk 4]
```diff
@@ -173,26 +173,30 @@ macro_rules! st_fields_enum {
...
-    "index_type", IndexType = 5,
-});
+st_fields_enum!(
+    enum StIndexFields {
+        "index_id", IndexId = 0,
+        "table_id", TableId = 1,
+        "index_name", IndexName = 2,
+
+        // TODO(jgilles): can a system table store an enum?
+        // What if we need to add new variants to the enum? Will that break
```
Comment (mamcx) [inline crates/core/src/db/datastore/system_tables.rs:195]: Yes, system tables can store any SATN type.

File: crates/core/src/db/datastore/system_tables.rs (hunk 2 above)
Comment (mamcx) [reply crates/core/src/db/datastore/system_tables.rs:179]: I agree to keep the prefix for the columns, that helps with debugging.

Comment (kazimuth) [reply crates/core/src/db/datastore/system_tables.rs:179]: Yeah and joins.

**Comment by kazimuth** [comment]
Closed in favor of #1697

**Pull request closed without merging**